Firestore/core/src/firebase/firestore/core/event_listener.h

```diff
@@ -87,9 +84,32 @@ class AsyncEventListener
   void Mute();
 
  private:
-  std::atomic<bool> muted_;
   std::shared_ptr<util::Executor> executor_;
   DelegateListener delegate_;
+
+  // A mutex that protects both muting the AsyncEventListener and also calling
+  // out to the delegate.
+  //
+  // Mute calls must be synchronized because users expect that when they call
+  // `ListenerRegistration::Remove` that they don't get notifications pretty
+  // much immediately upon return of that method. That is, we can't afford to
+  // wait for the `Remove` to be submitted through the `AsyncQueue`.
+  //
+  // The call to delegate_->OnEvent must also be protected in order to ensure
+  // that The `Firestore` instance isn't destroyed while we're calling out to
+  // user code. `Firestore::Dispose` (eventually) calls `Mute` on each listener
+  // and forcing `Mute` and `OnEvent` to be mutually exclusive avoids a race.
+  //
+  // This must be a recursive mutex because the `DelegateListener` may be user
+  // code, and that we must allow that user code to invoke
```

Comment:
Nit: the first `that` seems extraneous (`and we must allow that user code...` flows just as well).